### PR TITLE
Update: `no-use-before-define` checks invalid initializer

### LINKED ISSUE #4280: no-use-before-define issue
Hi, I'm new to eslint so there might be something that I'm missing, but I was expecting the rule "no-use-before-define" to pick up the following (contrived) syntax issue:

... | ....
----|----
version | v1.7.3
rule | no-use-before-define
code | see below
expected | output error
actual | no output

code:
```js
var a = a
```

eslintrc:
```json
{
    "extends": "eslint:recommended",
    "rules": {
         "no-use-before-define": 2
    }
}
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27773841-no-use-before-define-issue?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mysticatea**
Fixes #4280

And several invalid initialization were found.

### CODE REVIEW

File: lib/rules/no-use-before-define.js [hunk 1]
```diff
@@ -105,6 +153,10 @@ module.exports = function(context) {
      */
     function findVariablesInScope(scope) {
         scope.references.forEach(function(reference) {
+            if (reference.init) {
```
Comment (nzakas) [inline lib/rules/no-use-before-define.js]: Can you add a comment explaining why we exit here?

Comment (mysticatea) [inline lib/rules/no-use-before-define.js]: I'm thinking...

```js
var a = a;
```

In this case, escope makes 2 references. 

- One is the write reference of the first `a` for initialization.
- One is the read reference of the second `a`.

This rule does not need to handle the former, so it exits here.

Hmm, the comment might be "Does not need to handle references of initialization.".

Comment (mysticatea) [inline lib/rules/no-use-before-define.js]: I moved this to the next if statement.

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**